pkg/hive/hive.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package hive
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/hive/pb"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+const (
+	protocolName    = "hive"
+	protocolVersion = "1.0.0"
+	peersStreamName = "peers"
+	maxPO           = 7
+	messageTimeout  = 5 * time.Second // maximum allowed time for a message to be read or written.
+)
+
+type Service struct {
+	streamer          p2p.Streamer
+	connectionManager ConnectionManager
+	peerSuggester     PeerSuggester
+	addressFinder     AddressFinder
+	logger            logging.Logger
+}
+
+type Options struct {
+	Streamer          p2p.Streamer
+	ConnectionManager ConnectionManager
+	PeerSuggester     PeerSuggester
+	AddressFinder     AddressFinder
+	Logger            logging.Logger
+}
+
+func New(o Options) *Service {
+	return &Service{
+		streamer:          o.Streamer,
+		logger:            o.Logger,
+		connectionManager: o.ConnectionManager,
+		peerSuggester:     o.PeerSuggester,
+		addressFinder:     o.AddressFinder,
+	}
+}
+
+type BzzAddress struct {
+	Overlay, Underlay []byte
+}
+
+type ConnectionManager interface {
```

Comment:
I was thinking about the same name, could not decide between `Connecter` & `Connector`, but both somehow did not sound fully fight to me. I will switch to `Connecter`. 
I like the idea bout `p2p.Connecter`, the only question I had about this is do you guys think we should use string underlay or Multiaddress? If we use string, we could maybe put creation of Multiaddress inside the Connect method? Ltmwyt :)

Reply:
Ah yes, Connector, I do not know why I wrote Connecter.

I would rather go with specific Multiaddress type, then string, to avoid parsing and its error handling. We are parsing Multiaddress for bootnode cli flag, when hive receives the message with new peers, and in debugapi connect http handler, and from there, we can use concrete type everywhere internally. My view on the subject is to parse early and use concrete type as much as possible.

Reply:
i think connecter makes sense because it follows golang interface naming schemes (it sometimes looks strange but i don't have any particular problem with it). :+1: for multiaddr over strings